### PR TITLE
New report scan of licenses and violations

### PR DESCRIPTION
**Pull request by ns-dsinghal**
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [ ] This pull request is on the dev branch.
- [ ] I used gofmt for formatting the code before submitting the pull request.
-----

### CODE REVIEW

**Comment by ns-dsinghal** [comment]
> **CLA Assistant Lite bot:** Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you sign our [Contributor License Agreement](https://jfrog.com/cla/) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.
> 
> I have read the CLA Document and I hereby sign the CLA
> 
> You can retrigger this bot by commenting **recheck** in this Pull Request

I have read the CLA Document and I hereby sign the CLA

**Comment by ns-dsinghal** [comment]
recheck

**Comment by ns-dsinghal** [comment]
@talarian1 @eyalbe4 

Can you Please review and merge this PR

**Comment by ns-kbhat** [comment]
recehck

**Comment by ns-kbhat** [comment]
recheck

**Comment by ns-kbhat** [comment]
@talarian1 @eyalbe4 can you please take a look at this? The CLA acceptance is not being re-checked.

**Comment by ns-kbhat** [comment]
I have read the CLA Document and I hereby sign the CLA. Recheck

**Comment by ns-kbhat** [comment]
I have read the CLA Document and I hereby sign the CLA. recheck

**Review by yahavi — CHANGES_REQUESTED** [review]
Thanks for this contribution, @ns-dsinghal!

Could you please document this change in the README?
Regarding the CLA assistant - it expects a comment containing exactly "I have read the CLA Document and I hereby sign the CLA". Could you please add such a comment?

Thanks.

**Comment by ns-dsinghal** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by ns-dsinghal** [comment]
> Thanks for this contribution, @ns-dsinghal!
> 
> Could you please document this change in the README? Regarding the CLA assistant - it expects a comment containing exactly "I have read the CLA Document and I hereby sign the CLA". Could you please add such a comment?
> 
> Thanks.
Sure. CLA assistant worked now. let me add this change to README

**Comment by ns-dsinghal** [comment]
> Thanks for this contribution, @ns-dsinghal!
> 
> Could you please document this change in the README? Regarding the CLA assistant - it expects a comment containing exactly "I have read the CLA Document and I hereby sign the CLA". Could you please add such a comment?
> 
> Thanks.

README updated about this change

**Comment by ns-dsinghal** [comment]
> Thanks for this contribution, @ns-dsinghal!
> 
> Could you please document this change in the README? Regarding the CLA assistant - it expects a comment containing exactly "I have read the CLA Document and I hereby sign the CLA". Could you please add such a comment?
> 
> Thanks.

Done

File: README.md [hunk 1]
```diff
@@ -151,10 +151,10 @@
...
       - [Request Graph Scan](#request-graph-scan)
       - [Retrieve the Graph Scan Results](#retrieve-the-graph-scan-results)
-      - [Generate Vulnerabilities Report](#generate-vulnerabilities-report)
-      - [Get Vulnerabilities Report Details](#get-vulnerabilities-report-details)
-      - [Get Vulnerabilities Report Content](#get-vulnerabilities-report-content)
-      - [Delete Vulnerabilities Report](#delete-vulnerabilities-report)
+      - [Generate Vulnerabilities/Licences/Violations Report](#generate-vulnerabilities-licences-violations-report)
+      - [Get Vulnerabilities/Licenses/Violations Report Details](#get-vulnerabilities-licences-violations-report-details)
+      - [Get Vulnerabilities/Licenses/Violations Report Content](#get-vulnerabilities-licences-violations-report-content)
+      - [Delete Vulnerabilities/Licenses/Violations Report](#delete-vulnerabilities-licences-violations-report)
```
Comment (ns-kbhat) [inline README.md:157]: @ns-dsinghal the pre-existing links (above/below your change) goto sections that describe more about the api. I think similar sections needs to be added for the new lines you added.
Also I think the name of the links must be for a specific api (i.e. not include words separated by `/`)

Comment (ns-dsinghal) [reply README.md]: Fixed

**Review by yahavi — CHANGES_REQUESTED** [review]
@ns-dsinghal, 
Thanks again for this contribution!
According to the REST API documentation, the licenses filter can be for example:
```json
"filters": {
  "component": "*gav:component*",
  "artifact": "some://impacted*artifact",
  "unknown": false,
  "unrecognized": true,
  "license_names": [
      "Apache",
      "MIT"
      "AFL"
  ],
  "license_patterns": [
      "*Apache*",
      "The Apache*"
      "AFL*"
  ],
  "scan_date": {
      "start": "2020-06-29T12:22:16Z",
      "end": "2020-06-29T12:22:16Z"
  }
}
```
https://jfrog.com/help/r/jfrog-rest-apis/generate-due-diligence-report

A violation filter is even more complicated. It may contain either `security_filters` and `license_filters`:
```json
"filters": {
  "type": "security|license|operational_risk",
  "watch_names": [
    "NameOfWatch1",
    "NameOfWatch2"
  ],
  "watch_patterns": [
    "WildcardWatch*"
  ],
  "component": "*vulnerable:component*",
  "artifact": "some://impacted*artifact",
  "policy_names": [
    "NameOfPolicy"
  ],
  "severities": [
    "High",
    "Medium"
  ],
  "updated": {
    "start": "2020-01-02T15:00:00Z",
    "end": "2020-12-15T00:00:00Z"
  },
  "security_filters": {
    "cve": "CVE-2020-10693",
    "issue_id": "XRAY-87343",
    "cvss_score": {
      "min_score": 6.3,
      "max_score": 9
    },
    "summary_contains": "kernel",
    "has_remediation": false
  },
  "license_filters": {
    "unknown": false,
    "unrecognized": true,
    "license_names": [
      "Apache",
      "MIT",
      "AFL"
    ],
    "license_patterns": [
      "*Apache*",
      "AFL*"
    ]
  }
}
```
https://jfrog.com/help/r/jfrog-rest-apis/generate-violations-report

I understand your suggestion. Instead of using a general "Filter" implementation for generating licenses and violations reports, I propose the following updates:

1. Introduce a specific "VulnerabilitiesFilter" to filter vulnerabilities (instead of "Filter"). 
2. Create a new "LicensesFilter" that contains the essential functionalities for license filtering.
3. Introduce a "ViolationsFilter" that encompasses both the "VulnerabilitiesFilter" and "LicensesFilter" as the API requires.

These updates will provide more clarity and specific functionality for generating reports based on vulnerabilities, licenses, and violations.
Please let me know what you think.

**Comment by ns-kbhat** [comment]
@ns-dsinghal many thanks for your contributions!.

@yahavi please review. Looking forward to your review. If all is good, then looking forward to this available via a new release. Thanks!

File: xray/services/report.go [hunk 2]
```diff
@@ -83,22 +91,71 @@ type ReportCve struct {
...
+	Updated         DateTimeRange         `json:"updated,omitempty"`
+	SecurityFilters VulnerabilitiesFilter `json:"security_filters,omitempty"`
+	LicenseFilters  LicensesFilter        `json:"license_filters,omitempty"`
 }
 
 type CvssScore struct {
-	MinScore float32 `json:"min_score,omitempty"`
-	MaxScore float32 `json:"max_score,omitempty"`
+	MinScore float64 `json:"min_score,omitempty"`
+	MaxScore float64 `json:"max_score,omitempty"`
```
Comment (ns-kbhat) [inline xray/services/report.go:158]: @ns-dsinghal could you check if this change is really required? Lets avoid breaking back compat.

Comment (ns-dsinghal) [reply xray/services/report.go:158]: yes it is required. perviously it have only few parameter all the parameter can be empty. if this is not send it end up a bad query filter.

Comment (ns-kbhat) [reply xray/services/report.go:158]: @ns-dsinghal whatever you said, does that apply to the change in bits for `float` here (32->64)?

Comment (ns-dsinghal) [reply xray/services/report.go:158]: I have change to to float64 because other structure are also using float64 to hold CVE score

**Comment by ns-dsinghal** [comment]
@yahavi could you please review

File: xray/services/report.go (hunk 2 above)
Comment (ns-dsinghal) [reply xray/services/report.go]: restored to type float32

**Comment by yahavi** [comment]
@ns-dsinghal it looks like the static analysis failed. Would you like to take a look?

**Comment by ns-dsinghal** [comment]
recheck

**Comment by ns-dsinghal** [comment]
> @ns-dsinghal it looks like the static analysis failed. Would you like to take a look?

I have fixed the issue can you re-run static analysis job

**Comment by ns-kbhat** [comment]
> > @ns-dsinghal it looks like the static analysis failed. Would you like to take a look?
> 
> I have fixed the issue can you re-run static analysis job

@yahavi need your attention here... looking forward to get this merged, thanks!

**Comment by ns-kbhat** [comment]
> > > @ns-dsinghal it looks like the static analysis failed. Would you like to take a look?
> > 
> > 
> > I have fixed the issue can you re-run static analysis job
> 
> @yahavi need your attention here... looking forward to get this merged, thanks!

@yahavi @eyalbe4 please review this.

**Comment by yahavi** [comment]
@ns-dsinghal it looks like the static analysis is still failing

**Comment by ns-kbhat** [comment]
@ns-dsinghal it appears adding the `safe-to-test` label initiates the static analysis.

Please try that label on your changes and see if that helps get early feedback.

**Comment by ns-dsinghal** [comment]
> @ns-dsinghal it appears adding the `safe-to-test` label initiates the static analysis.
> 
> Please try that label on your changes and see if that helps get early feedback.

label is getting removed by github-actions bot 

I have run static check locally using same command it give no error after my changes. 

@yahavi you can review now

File: go.sum [hunk 3]
```diff
@@ -13,7 +13,8 @@ github.com/armon/go-socks5 v0.0.0-20160902184237-e75332964ef5 h1:0CwZNZbxp69SHPd
 github.com/bradleyjkemp/cupaloy/v2 v2.8.0 h1:any4BmKE+jGIaMpnU8YgH/I2LPiLBufr6oMMlVBbn9M=
 github.com/buger/jsonparser v1.1.1 h1:2PnMjfWD7wBILjqQbt530v576A/cAbQvEW9gGIpYMUs=
 github.com/buger/jsonparser v1.1.1/go.mod h1:6RYKKt7H4d4+iWqouImQ9R2FZql3VbhNgx27UK13J/0=
-github.com/bwesterb/go-ristretto v1.2.3/go.mod h1:fUIoIZaG73pV5biE2Blr2xEzDoMj7NFEuV9ekS419A0=
+github.com/bwesterb/go-ristretto v1.2.0/go.mod h1:fUIoIZaG73pV5biE2Blr2xEzDoMj7NFEuV9ekS419A0=
```
Comment (ns-kbhat) [inline go.sum:16]: how did this change get applied?

File: xray/manager.go [hunk 4]
```diff
@@ -143,12 +143,26 @@ func (sm *XrayServicesManager) BuildScan(params services.XrayBuildParams, includ
...
 
+// GenerateLicensesReport returns a Xray report response of the requested report
+func (sm *XrayServicesManager) GenerateLicensesReport(params services.LicensesReportRequestParams) (resp *services.ReportResponse, err error) {
+	reportService := services.NewReportService(sm.client)
+	reportService.XrayDetails = sm.config.GetServiceDetails()
+	return reportService.Licenses(params)
+}
+
+// GenerateVoilationsReport returns a Xray report response of the requested report
+func (sm *XrayServicesManager) GenerateVoilationsReport(params services.ViolationsReportRequestParams) (resp *services.ReportResponse, err error) {
```
Comment (yahavi) [inline xray/manager.go:160]:
```suggestion
func (sm *XrayServicesManager) GenerateViolationsReport(params services.ViolationsReportRequestParams) (resp *services.ReportResponse, err error) {
```

File: xray/services/report.go [hunk 5]
```diff
@@ -122,12 +179,27 @@ func NewReportService(client *jfroghttpclient.JfrogHttpClient) *ReportService {
...
+	return rs.requestReport(req, Vulnerabilities)
+}
+
+// Licenses requests a new Xray scan for licenses
+func (rs *ReportService) Licenses(req LicensesReportRequestParams) (*ReportResponse, error) {
+	return rs.requestReport(req, Licenses)
+}
+
+// Violations requests a new Xray scan for violations
+func (rs *ReportService) Violations(req ViolationsReportRequestParams) (*ReportResponse, error) {
```
Comment (yahavi) [inline xray/services/report.go:192]: Looks like the new APIs are not covered with tests. We are unable to add new code without tests.
Let's add some tests "reportLicenses" and "reportViolaions" in xrayreport_test.go to cover this new logic.

**Review by yahavi — CHANGES_REQUESTED** [review]

File: xray/services/report.go (hunk 5 above)
Comment (ns-dsinghal) [reply xray/services/report.go:203]: @yahavi I have modified existing test testfile and added new test for license xray report api.

**Comment by ns-kbhat** [comment]
@yahavi @eyalbe4 could you please review and merge this?

**Comment by yahavi** [comment]
@ns-kbhat @ns-dsinghal - could you please fix the static code analysis issues?

**Comment by ns-kbhat** [comment]
> > @ns-kbhat @ns-dsinghal - could you please fix the static code analysis issues?
> 
> Fixed and checked locally.

cc @yahavi @eyalbe4

**Comment by ns-kbhat** [comment]
@yahavi what is the next step here?

**Comment by yahavi** [comment]
Thanks for your contribution, @ns-dsinghal & @ns-kbhat!
I merged the PR. 
Please inform us if everything is functioning correctly.

**Comment by ns-kbhat** [comment]
@ns-dsinghal did all the work, so kudos to him.

@yahavi when can we expect this change to be rolled into a release?

**Pull request merged**